xdsl/ir/core.py

```diff
@@ -2011,6 +2011,23 @@ def parent_region(self) -> Region | None:
             else None
         )
 
+    def find_ancestor_block_in_region(self, block: Block) -> Block | None:
+        """
+        Returns 'block' if 'block' lies in this region, or otherwise finds
+        the ancestor of 'block' that lies in this region.
+
+        Returns None if no ancestor block that lies in this region is found.
+        """
+        curr_block = block
+        while curr_block.parent_region() != self:
+            parent_op = curr_block.parent_op()
+            if not parent_op or not parent_op.parent_block():
+                return None
+            curr_block = parent_op.parent_block()
+            assert isinstance(curr_block, Block)
```

Comment:
```suggestion
            curr_block = parent_op.parent_block()
            if curr_block is not None:
                return None
```
Should be more concise

Reply:
Oh sorry, replace `is not None` with `is None` in my code!

Reply:
Also, a check on `parent_op` needs to be kept, otherwise it could be `None` when calling `parent_block()`. I think the code is in its minimal form at the moment to keep Pyright happy. @math-fehr please, let me know if you agree and then I'll merge!

Reply:
I have a counter proposal that also avoids isinstance:
``` python
        curr_block = block
        while (
            curr_block is not None
            and (parent_region := curr_block.parent_region()) is not None
        ):
            if parent_region is self:
                return curr_block
            curr_block = parent_region.parent_block()
```

Reply:
I'm fine with your version or mine tbh

Reply:
Thank you @superlopuh, that version seems correct too, but @math-fehr seems simpler and passes the tests. Let's go with that one, I think :)

Reply:
Oh actually this works too, fun puzzle:
``` python
        curr_block = block
        while (parent_region := curr_block.parent_region()) is not None:
            if (
                parent_region is self
                or (curr_block := parent_region.parent_block()) is None
            ):
                return curr_block
```

Reply:
I just got what Mathieu was going for, final offer:
``` python
        curr_block = block
        while curr_block is not None and curr_block.parent_region() is not self:
            curr_block = curr_block.parent_block()
        return curr_block
```